### PR TITLE
Move test framework code into distinct framework classes

### PR DESCRIPTION
**Pull request by happz**
The implementation of various framework-specific steps is now wrapped by their own plugin construct.

### CODE REVIEW

**Comment by martinhoyer** [comment]
> happz force-pushed the test-frameworks branch from f91daf3 to 29e3536 8 minutes ago

guess I should have waited 20 minutes before [rebasing](https://github.com/martinhoyer/tmt/commit/c06a1cd65971fcf72a309d1ebb14d79c9f3be8d0) it myself lol

File: tmt/frameworks/__init__.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+FrameworkClass = Type['ExecuteFramework']
+
+
+_FRAMEWORK_PLUGIN_REGISTRY: tmt.plugins.PluginRegistry[FrameworkClass] = \
+    tmt.plugins.PluginRegistry()
+
+
+def provides_framework(framework: str) -> Callable[[FrameworkClass], FrameworkClass]:
+    """
+    A decorator for registering export format.
```
Comment (psss) [inline tmt/frameworks/__init__.py:23]: A copy paste typo? Should mention `registering execute framework` instead?

File: tmt/frameworks/__init__.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            plugin_id=framework,
+            plugin=framework_cls,
+            logger=tmt.log.Logger.get_bootstrap_logger())
+
+        return framework_cls
+
+    return _provides_framework
+
+
+class ExecuteFramework:
```
Comment (psss) [inline tmt/frameworks/__init__.py:39]: Please, add a docstring.

File: tmt/frameworks/__init__.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        """
+        Provide additional environment variables for the test.
+
+        :param parent: ``execute`` plugin managing the test.
+        :param test: a test that would be executed.
+        :param guest: a guest on which the test would run.
+        :param logger: to use for logging.
+        :returns: environment variables to expose for the test. Variables
+            would be added on top of any variables the plugin, test or plan
+            might already collected.
```
Comment (psss) [inline tmt/frameworks/__init__.py:56]:
```suggestion
            might have already collected.
```

File: tmt/frameworks/__init__.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+            guest: 'Guest',
+            logger: tmt.log.Logger) -> List[str]:
+        """
+        Provide additional options for pulling test data directory.
+
+        :param parent: ``execute`` plugin managing the test.
+        :param test: a test that would be executed.
+        :param guest: a guest on which the test would run.
+        :param logger: to use for logging.
+        :returns: additional options for a ``rsync`` tmt would use to pull
```
Comment (psss) [inline tmt/frameworks/__init__.py:96]:
```suggestion
        :returns: additional options for the ``rsync`` command tmt would use to pull
```

**Review by psss — CHANGES_REQUESTED** [review]
Thanks much for reimplementing the `framework` with plugins. Looks really nice! Added just a few nitpick comments.

File: tmt/frameworks/__init__.py (hunk 1 above)
Comment (happz) [reply tmt/frameworks/__init__.py]: Fixed in 76bc42ba.

File: tmt/frameworks/__init__.py (hunk 2 above)
Comment (happz) [reply tmt/frameworks/__init__.py]: Fixed in 76bc42ba.

File: tmt/frameworks/__init__.py (hunk 3 above)
Comment (happz) [reply tmt/frameworks/__init__.py]: Fixed in 76bc42ba.

File: tmt/frameworks/__init__.py (hunk 4 above)
Comment (happz) [reply tmt/frameworks/__init__.py]: Fixed in 76bc42ba.

**Review by psss — APPROVED** [review]
Thanks for addressing all comments, now looks good.

**Pull request merged**